sklearn/metric_learning/nca.py

```diff
@@ -0,0 +1,216 @@
+from ..base import BaseEstimator, TransformerMixin
+
+import numpy as np
+import scipy as sp
+import scipy.misc
+import scipy.optimize
+import time
+
+import sys
+
+def nca_vectorized_oracle(X, y, params):
+    dX = X[:, None, :] - X[None, :, :]  # n_samples x n_samples x n_comp
+    outer = dX[:, :, None, :] * dX[:, :, :, None]  # n_samples x n_samples x n_comp x n_comp
+
+    n_samples, n_features = X.shape
+    n_components = params.n_components
+
+    def oracle(L):
+        L = L.reshape((n_components, n_features))
+
+        Lx = np.dot(X, L.T)  # n_samples x n_comp
+        assert Lx.shape == (n_samples, n_components)
+
+        A = Lx[None, :, :] - Lx[:, None, :]  # n_samples x n_samples x n_comp
+        assert A.shape == (n_samples, n_samples, n_components)
+
+        logp = -(A*A).sum(axis=2)  # n_samples x n_samples
+        np.fill_diagonal(logp, -np.inf)
+        logp -= sp.misc.logsumexp(logp, axis=1)[:, None]
+        assert logp.shape == (n_samples, n_samples)
+
+        p = np.exp(logp)  # n_samples x n_samples
+        assert p.shape == (n_samples, n_samples)
+
+        class_neighbours = y[:, None] == y[None, :]
+
+        p_i = (p * class_neighbours).sum(axis=1)
+
+        if params.loss == 'l1':
+            grad = np.sum(np.sum(p[:, :, None, None] * outer, axis=0) * p_i[:, None, None], axis=0)
+        else:
+            grad = np.sum(np.sum(p[:, :, None, None] * outer, axis=0), axis=0)
+
+        assert grad.shape == (n_features, n_features)
+
+        if params.loss == 'l1':
+            snd = np.sum(class_neighbours[:, :, None, None] * p[:, :, None, None] * outer, axis=(0, 1))
+        else:
+            snd = np.sum(class_neighbours[:, :, None, None] * p[:, :, None, None] * outer, axis=0) / p_i[:, None, None]
+            snd = np.sum(snd, axis=0)
+
+        assert snd.shape == (n_features, n_features)
+        grad -= snd
+
+        fnc = p_i.sum() if params.loss == 'l1' else np.log(p_i).sum()
+
+        grad = 2 * np.dot(L, grad)
+
+        return [-fnc, -grad.flatten()]
+
+    return oracle
+
+
+def nca_semivectorized_oracle(X, y, params, threshold=None):
+    n_samples, n_features = X.shape
+    n_components = params.n_components
+
+    if threshold is None:
+        threshold = 1e-3 / np.abs(X).max() ** 2
+
+    def oracle(L):
+        L = L.reshape((n_components, n_features))
+
+        Lx = np.dot(X, L.T)  # n_samples x n_comp
+        assert Lx.shape == (n_samples, n_components)
+
+        grad = np.zeros((n_features, n_features))
+        fnc = 0
+
+        for i in range(n_samples):
+            A = Lx[i, :] - Lx  # n_samples x n_comp
+            assert A.shape == (n_samples, n_components)
+
+            logp = -(A*A).sum(axis=1)  # n_samples
+            logp[i] = -np.inf
+            logp -= sp.misc.logsumexp(logp)
+            assert logp.shape == (n_samples, )
+
+            p = np.exp(logp)  # n_samples
```

Comment:
This bothers me. Sometimes I get various numerical warnings (underflows). Should I keep probabilities in log-domain and use log-sum-exp trick for sums?